api/src/main/java/io/strimzi/api/kafka/model/listener/v2/ArrayOrObjectKafkaListeners.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.api.kafka.model.listener.v2;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.core.ObjectCodec;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.strimzi.api.kafka.model.Constants;
+import io.strimzi.api.kafka.model.listener.KafkaListeners;
+import io.strimzi.crdgenerator.annotations.Alternation;
+import io.strimzi.crdgenerator.annotations.Alternative;
+import io.sundr.builder.annotations.Buildable;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.List;
+
+@Buildable(
+        editableEnabled = false,
+        builderPackage = Constants.FABRIC8_KUBERNETES_API
+)
+@JsonDeserialize(using = ArrayOrObjectKafkaListeners.Deserializer.class)
+@JsonSerialize(using = ArrayOrObjectKafkaListeners.Serializer.class)
+@Alternation
+public class ArrayOrObjectKafkaListeners implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    private final List<GenericKafkaListener> listValue;
+    private final KafkaListeners objectValue;
+
+    public ArrayOrObjectKafkaListeners(List<GenericKafkaListener> listValue)   {
+        this.listValue = listValue;
+        this.objectValue = null;
+    }
+
+    public ArrayOrObjectKafkaListeners(KafkaListeners objectValue)   {
+        this.listValue = null;
+        this.objectValue = objectValue;
+    }
+
+    @Alternative()
+    public List<GenericKafkaListener> getListValue() {
+        return listValue;
+    }
+
+    @Alternative()
+    public KafkaListeners getObjectValue() {
+        return objectValue;
+    }
+
+    /**
+     * Convenience method which returns either the new listener format if set, or converted old format.
+     *
+     * @return  List of new listeners
+     */
+    public List<GenericKafkaListener> newOrConverted()  {
+        if (listValue != null)  {
+            return listValue;
+        } else {
+            return ListenersConvertor.convertToNewFormat(objectValue);
+        }
+    }
+
+    public static class Serializer extends JsonSerializer<ArrayOrObjectKafkaListeners> {
+        @Override
+        public void serialize(ArrayOrObjectKafkaListeners value, JsonGenerator generator, SerializerProvider provider) throws IOException {
```

Comment:
```suggestion
        public void serialize(ArrayOrObjectKafkaListeners value, JsonGenerator generator, SerializerProvider provider) throws IOException {
        if (value == null) {
            generator.writeNull();
            return;
        }
                if (value.listValue != null)    {
                    generator.writeObject(value.listValue);
                    return;
                } 
                if (value.objectValue != null)  {
                    generator.writeObject(value.objectValue);
                    return;
                }
                generator.writeNull();
                return;
               
        }
```

Reply:
I think this makes it much harder to read. What are the advantages?

Reply:
Personally I find nested ifs harder to read than an explicit list of priorities

This presents more like a checklist,

if val == null write null
if listValue != nul then return that.
Especially with things like trace entry/exit enabled it makes it easier to reasn where your code reaches and when it is early exiting.

If/elsiif isn't bad, but it leads to less maintainable code as it is added to, from my experience, not sure if other maintainers such as @ppatierno might have opinon on this.
Obviously it is just an issue of styling, but it can matter.

Reply:
Ok, I changed it a bit. Hope you find it more readable now.